### DISCUSSION TITLE
Recover samples from an unfinished file

### DISCUSSION THREAD

**Discussion by tiagoporsch** [ihedvall/mdflib#89] [Q&A] [answered]
Hello, thank you for the work put into developing this library.

I am using it to create a new MDF file and incrementally save samples to it, just like [here](https://github.com/ihedvall/mdflib/blob/e300448b21f6e10b8ca846e33e3f060fcd7dcad5/mdflib_test/src/testwrite.cpp#L189).
If, for whatever reason, my program suddenly stops executing and never recovers, none of the resulting channels in the MDF have any cycles saved.
However, I can see that the file size gets incrementally bigger while the program is running, so some data is being saved to it.
Is there a way to recover this data?

Thank you in advance.

**Comment by tiagoporsch** [comment]
I found a probable solution and created a pull request #91. Not sure if it's the best, but it sure works!

**Reply by ihedvall** [reply, marked as answer]
Thanks for finding and fix this bug in MDF 3 writers. This is actually a bug as your code change, so to say exists, in the MDF 4 writer SaveQueue() function.

Note that some loggers as from the CSS Electronics company, just append bytes at the end of the file without updating any of the internal counters. Instead the mark the file as unfinished in the initial ID block. This causes problems for most MDF readers as it appears not exist any samples.

The MDF 4 reader check this and finalize the files by updating the counters (it's many in MDF 4). There is a time penalty for doing this as you have to read the data block twice.
